scalameta/semanticdb-scalac/src/main/scala/scala/meta/internal/semanticdb/DocumentOps.scala

```diff
@@ -175,271 +177,274 @@ trait DocumentOps { self: DatabaseOps =>
+                success(morePrecisePos, _.copy(targs = Some(syntax)))
+              case ApplySelect(select @ g.Select(qual, nme)) if isSyntheticName(select) =>
+                val pos = qual.pos.withStart(qual.pos.end).toMeta
+                val symbol = select.symbol.toSemantic
+                val name = nme.decoded
+                val names = List(SyntheticRange(0, name.length, symbol))
+                val syntax = S(".") + S(nme.decoded, names)
+                success(pos, _.copy(select = Some(syntax)))
+              case _ =>
+              // do nothing
             }
           }
-          traverser.traverse(unit.body)
+
+          override def traverse(gtree: g.Tree): Unit = {
+            if (isVisited(gtree)) return else isVisited += gtree
+            gtree.attachments.all.foreach {
+              case att: g.analyzer.MacroExpansionAttachment =>
+                traverse(att.expandee)
+              case _ =>
+            }
+            gtree match {
+              case ConstfoldOf(original) =>
+                traverse(original)
+              case ClassOf(original) =>
+                traverse(original)
+              case NewArrayOf(original) =>
+                traverse(original)
+              case SingletonTypeTreeOf(original) =>
+                traverse(original)
+              case CompoundTypeTreeOf(original) =>
+                traverse(original)
+              case ExistentialTypeTreeOf(original) =>
+                traverse(original)
+              case AnnotatedOf(original) =>
+                traverse(original)
+              case SelfTypeOf(original) =>
+                traverse(original)
+              case SelectOf(original) =>
+                traverse(original)
+              case g.Function(params, body) if params.exists(_.name.decoded.startsWith("x$")) =>
+                traverse(body)
+              case gtree: g.TypeTree if gtree.original != null =>
+                traverse(gtree.original)
+              case gtree: g.TypeTreeWithDeferredRefCheck =>
+                traverse(gtree.check())
+              case gtree: g.MemberDef =>
+                gtree.symbol.annotations.map(ann => traverse(ann.original))
+                tryFindMtree(gtree)
+              case _: g.Apply | _: g.TypeApply =>
+                tryFindInferred(gtree)
+              case select: g.Select if isSyntheticName(select) =>
+                tryFindMtree(select.qualifier)
+                tryFindInferred(select)
+              case _ =>
+                tryFindMtree(gtree)
+            }
+            super.traverse(gtree)
+          }
         }
+        traverser.traverse(unit.body)
+      }
 
-        val input = unit.source.toInput
+      val input = unit.source.toInput
 
-        val synthetics = inferred.toIterator.map {
-          case (pos, inferred) => inferred.toSynthetic(input, pos)
-        }
+      unit.body.appendMetadata("semanticdbMstarts" -> mstarts.mapValues(_.pos).toMap) // used in MessagesOps
```

Comment:
`m.Position` holds a reference to `m.Input`. I think we can keep only the (start, end) ints to lower memory usage.

Reply:
I can make this change.